### ISSUE TITLE
Invalid graph after running bert_tf optimizer

### ISSUE THREAD

**Issue by brosenfi** [microsoft/onnxruntime#7446]
**Describe the bug**
The graph produced by using onnxruntime/onnxruntime/python/tools/transformers/onnx_model_bert_tf.py on a tf2onnx converted bert classifier model is invalid and fails with the following error during a load:

> Load model from <onnx file location> failed:This is an invalid model. Node () input arg (Squeeze_3d_to_2d_mask_output) does not have type information set by parent node.

**Urgency**
none

**System information**
- System OS RHEL 7, Container OS ubuntu 20.04 LTS
- onnxruntime-gpu installed using the pypi wheel (binary)
- ONNX Runtime version: 1.7.0
- Python version: 3.8
- CUDA/cuDNN version: 11.2
- GPU model and memory: Tesla V100 16GB

**To Reproduce**
Train a fine-tuned BERT large model via the traditional bert/run_classifier.py - produce a saved model with the standard raw serving interface (input_ids, input_mask, segment_ids, etc.).
Export that model using tf2onnx
Run the "bert_tf" optimizer with a minimum optimization level of 1 = basic
Resulting graph cannot be successfully loaded and executed by onnxruntime.

**Expected behavior**
I expected after the optimizer completed successfully that a valid graph that could be loaded and executed would be produced.

**Additional context**
This is a Tensorflow based BERT large model (trained with a variant of the traditional bert/run_classifier.py) - this is NOT a converted huggingface / transformers model.

fuse_attention() is the logical culprit, and I'm wondering if a recent change has caused this - specifically #4fd9fef9ee04c0844d679e81264779402cfa445c - @liuziyue ?

**Comment by mrry** [comment]
@liuziyue Could you please take a look and confirm whether 4fd9fef9ee04c0844d679e81264779402cfa445c has changed the behavior here? Thanks!

**Comment by tianleiwu** [comment]
@brosenfi, could you share an onnx model that could reproduce the issue?

**Comment by brosenfi** [comment]
Hi @tianleiwu - I'm going to have to confirm with others at my company whether it's allowed (may take a few days).  Also this is a huge file of course with all the weights (~ 1.3GB) - is there metadata that I could produce instead of uploading the huge file that would be useful for reproducing?

**Comment by tianleiwu** [comment]
@brosenfi, you can export the onnx model with external data format (use --large_model in tf2onnx). In that format, graph and weights are stored in separated files. The graph (.onnx file) is very small and I think it is enough for investigation, and you need not share the weights.

**Comment by brosenfi** [comment]
Thanks for the tip @tianleiwu - I tried doing this a little bit ago but ran into a couple of problems - first off the input we use to tf2onnx is not a saved model, which apparently is required - we are instead freezing the saved model first with a call to tf.graph_util.convert_variables_to_constants .  The other issue is that is requires a TF 2.x model - ours is TF 1.x .

However - I still can convert the graph to "external data" using onnx itself (thankfully this is now available as of version 1.9.0 = https://github.com/onnx/onnx/blob/master/docs/ExternalData.md ).  I will get you this file shortly.

**Comment by brosenfi** [comment]
Here it is... https://github.com/brosenfi/onnx-optimizer-issue/blob/master/model.onnx

**Comment by tianleiwu** [comment]
Have two PRs related to the model:
(1) Fix symbolic shape inference for Unsqueeze of opset 13: https://github.com/microsoft/onnxruntime/pull/7555
(2) Catch exception in shape inference and clean up prune graph: https://github.com/microsoft/onnxruntime/pull/7560

After these changes, optimizer could run on the model. However, attention can not be fused since mask is different. A possible fix is to add a new pattern for mask subgraph like https://github.com/microsoft/onnxruntime/blob/d7b84e67fe0be3be9d680331005e59b278511f13/onnxruntime/python/tools/transformers/onnx_model_bert_tf.py#L45-L49.
